src/workerd/api/sockets.h

```diff
@@ -36,19 +36,19 @@ class PipelinedAsyncIoStream final : public kj::AsyncIoStream, public kj::Refcou
 
 private:
   template <typename T>
-  kj::Promise<T> thenOrRunNow(std::function<kj::Promise<T> (kj::Own<kj::AsyncIoStream> *)> f) {
+  kj::Promise<T> thenOrRunNow(std::function<kj::Promise<T> (kj::AsyncIoStream&)> f) {
     // Either calls `then` on the `inner` promise with the `f` callback, or calls `f` on an already
     // stored stream (if the `inner` promise completed).
     KJ_IF_MAYBE(e, error) {
       kj::throwRecoverableException(kj::mv(*e));
     }
 
     KJ_IF_MAYBE(s, completedInner) {
-      return f(s);
+      return f(*s->get());
```

Comment:
This isn't a new issue here but worth pointing out: Since the `PipelinedAsyncIoStream` that actually owns `s` here is refcounted, it probably makes sense to attach a ref to this promise when returning it to ensure that the ref remains valid.